src/vs/workbench/services/runtimeSession/common/runtimeSession.ts

```diff
@@ -2019,7 +2046,129 @@ export class RuntimeSessionService extends Disposable implements IRuntimeSession
 		this._updateService.updateActiveLanguages([...languages]);
 	}
 
+	/**
+	 * Updates the URI of a notebook session to maintain session continuity when
+	 * a notebook is saved under a new URI.
+	 *
+	 * This is a crucial operation during the Untitled → Saved file transition, as it:
+	 * 1. Preserves all runtime state (variables, execution context, kernel connections)
+	 * 2. Updates internal mappings to reflect the new URI
+	 * 3. Notifies dependent components about the change (via the onDidUpdateNotebookSessionUri event)
+	 *
+	 * The implementation carefully orders operations to maintain state consistency even if
+	 * an error occurs during the update process.
+	 *
+	 * @param oldUri The original URI of the notebook (typically an untitled:// URI)
+	 * @param newUri The new URI of the notebook (typically a file:// URI after saving)
+	 * @returns The session ID of the updated session, or undefined if no update occurred
+	 */
+	updateNotebookSessionUri(oldUri: URI, newUri: URI): string | undefined {
+		// Validate inputs - both URIs must be defined
+		// Why: We need valid URIs to perform lookups and avoid system errors
+		if (!oldUri || !newUri) {
+			this._logService.error('updateNotebookSessionUri called with invalid URIs', { oldUri, newUri });
+			return undefined;
+		}
```

Comment:
If they can be undefined, should the types reflect that?

Reply:
They cant be, this was left over from a previous implementation. Good catch!